### ISSUE TITLE
Refactor meta_features in StackingCVRegressor (and the other Stacking classes) so that users can obtain these meta_features via a method call 

### ISSUE THREAD

**Issue by tianke0711** [rasbt/mlxtend#285]
I found some code that using the StackingCVRegressor to get the meta_features, pred_meta_features.  But I searched that there is no such kind of attributes in the document. Could you guys tell me more about these attributes.

**Comment by rasbt** [comment]
Hi, there,

there's currently no function or attribute for that. But if you look at the code in the predict function:

https://github.com/rasbt/mlxtend/blob/66394e0a93fb2420a4338875507cc5f69ff73b74/mlxtend/regressor/stacking_cv_regression.py#L152-L159

you should be able to get the meta features via


        np.column_stack([ regr.predict(X) for regr in your_regressor_instance.regr_ ]) 

I am curious, where did you see the code involving `f2 = stacking_mode.pred_meta_features`? Was this in the documenation or is that a modified version of mlxtend's `StackingCVRegressor`?

**Comment by tianke0711** [comment]
@rasbt, the code used the import code to use the  from mlxtend.regressor import StackingCVRegressor. There is meta_features and pred_meta_features function in the code. But I searched in the document, there is no such kind of function.

**Comment by rasbt** [comment]
Hm, that's strange. No idea what's going on. The code does not have such functions, e.g., see: https://github.com/rasbt/mlxtend/blob/master/mlxtend/regressor/stacking_cv_regression.py

Which version of mlxtend are you using?

**Comment by tianke0711** [comment]
@rasbt  you are right, that's new version of mlxtend!please access to this link:https://github.com/takashioya/mlxtend/tree/master/mlxtend/regressor

**Comment by rasbt** [comment]
Thanks for the note. To avoid confusion, that's not a new version of mlxtend -- i.e., when you run `pip install mlxtend` or install mlxtend via conda-forge, you will get the version from the repo here. The repo you linked is a fork of the last development version that has been refactored a bit to have this function:

<img width="803" alt="screen shot 2017-11-21 at 9 57 00 pm" src="https://user-images.githubusercontent.com/5618407/33107816-f2641048-cf06-11e7-9b41-fef2f298ffe3.png">

I think it's definitely useful to have that function separated. I would welcome a PR @takashioya if you have time some time :)

**Comment by rasbt** [comment]
Glad that the issue seems to be resolved. However, let me reopen that issue as a reminder to add this feature/modification in future.

**Comment by tianke0711** [comment]
sorry, I missed to click the close function, you can reopen it, if you want.

**Comment by takashioya** [comment]
hi, rasbt. 
I think It’s not diffucult to add function that extracts meta_features.

Like this:

```python
def _predict_meta_features(self, X):
    return np.column_stack([regr.predict(X) for regr in self.regr_])
```

However, it’s a bit difficult to get ‘meta_features’ from training data, because ‘_pred_meta_features’ doesn’t work.

Like this:

```python
model = StackingCVRegressor()
model.fit(X, y)
# clearly it's a leakage and meaningless.
X_meta_features = _predict_meta_features(X)
```

So, I think I should add a parameter like ‘return_meta_features’(bool) for ‘fit’ or a method like ‘fit_transform’.

Like this(This is almost same as ‘fit’, so adding parameters may be better.):

```python
def fit_transform(self, X, y, groups=None):
    self.regr_ = [clone(x) for x in self.regressors]
    self.meta_regr_ = clone(self.meta_regressor)

    kfold = check_cv(self.cv, y)
    if isinstance(self.cv, int):
        kfold.shuffle = self.shuffle

    meta_features = np.zeros((X.shape[0], len(self.regressors)))

    for i, regr in enumerate(self.regressors):
        for train_idx, holdout_idx in kfold.split(X, y, groups):
            instance = clone(regr)
            instance.fit(X[train_idx], y[train_idx])
            y_pred = instance.predict(X[holdout_idx])
            meta_features[holdout_idx, i] = y_pred

    if self.use_features_in_secondary:
        return np.hstack((X, meta_features))
    else:
        return meta_features
```

How do you think I should do about that?

**Comment by rasbt** [comment]
Thanks for looking into this @takashioya !

Regarding your comment:

> However, it’s a bit difficult to get ‘meta_features’ from training data, because > ‘_predict_meta_features’ doesn’t work.

> Example:

> model = StackingCVRegressor()
> model.fit(X, y)
>  #clearly it's a leakage and meaningless.
> X_meta_features = model._predict_meta_features(X)

I am wondering why this wouldn't work and what do you mean by leakage exactly? Modifying the "fit" method would be another option like you said. However, the issue with that is that it would be mixing the model fitting with a prediction step, which may be a bit more confusing for users? Also, if someone wants to predict the "meta features" for the test set, for example, it would require fitting the regressor again (on the test set).

**Comment by takashioya** [comment]
@rasbt
Thank you for reply.
I apologize to you for my poor explanation (and English).

I think, there are 2 types of 'meta_features' in Stacking (which is using CV).
The first one is 'meta features' for the test set, and the second one is 'meta_features' for the training set.

In order to clarify the difference, I took code from 'mlxtend/regressor/stacking_cv_regression.py'.

In Code_1, I showed the 'meta_features' for the test set, 
In Code_2, I showed the 'meta_features' for the training set.

Code_1:
```python
def predict(self, X):
    #This is the meta features for the test set
    meta_features = np.column_stack([
            regr.predict(X) for regr in self.regr_
        ])
    if self.use_features_in_secondary:
        return self.meta_regr_.predict(np.hstack((X, meta_features)))
    else:
        return self.meta_regr_.predict(meta_features)
```

Code_2:
```python
def fit(self, X, y, groups=None):
    kfold = check_cv(self.cv, y)
    if isinstance(self.cv, int):
        kfold.shuffle = self.shuffle

    #This is the meta features for the training set
    meta_features = np.zeros((X.shape[0], len(self.regressors)))

    for i, regr in enumerate(self.regressors):

        #Here, unlike 'predict', CV is used to get meta features for training set
        for train_idx, holdout_idx in kfold.split(X, y, groups):
            instance = clone(regr)
            instance.fit(X[train_idx], y[train_idx])
            y_pred = instance.predict(X[holdout_idx])
            meta_features[holdout_idx, i] = y_pred

    if self.use_features_in_secondary:
        self.meta_regr_.fit(np.hstack((X, meta_features)), y)
    else:
        self.meta_regr_.fit(meta_features, y)

    for regr in self.regr_:
        regr.fit(X, y)

    return self
```

In the present implementation, users can get neither of the 'meta_features'.
I think both of them are very useful. So, if possible, I would like to add function that extracts not only the 'meta features' for the test set, but also the 'meta features' for the training set.

To get the 'meta_features' for the test set, We don't need to modify the 'fit' method. (it's also very confusing as you say.)
So, Just adding the codes like 'Example 1' (shown in my last comment) should be enough.

However, using codes like 'Example 1', we can't get 'meta features' for training set, because in '_predict_meta_features' we don't use models trained using CV, but use models trained on all data.
(That's why I said 'leakage'. Leakage doesn't happen in 'fit' thanks to CV.)
So, maybe we need to modify 'fit' or add another method to extract 'meta features' for training set.

**Comment by rasbt** [comment]
No worries, and I see what you mean now -- I was only thinking of the "test" meta_features previously, which is why I maybe got confused. But sure, having options for both would be fine :).

So regarding the "test" meta_feature, i.e., the easier case, it would be fine to do the following as you suggested:

```python
def _predict_meta_features(self, X):
    return np.column_stack([regr.predict(X) for regr in self.regr_])
```

However, we should make this a "public" method; that is, renaming it to `def predict_meta_features(self, X)`.

Now, thinking about it a bit, I think it would be good to have 2 methods: the one above and a second one for the training meta features so that people can compute the train meta_features (e.g., later) without calling fit again. Also, per scikit-learn convention, it's maybe confusing if `fit` returns data. 

My preferred solution would look like this though: `predict_meta_features(self, X=None)`. And if `X=None` then the training meta features are returned, and otherwise, the meta features for X are returned. 

I am not 100% positive on that, but the CV as it is currently implemented -- regarding the shuffle argument -- should be deterministic and always return the same shuffling order. (Otherwise we might have to set the random seed somehow manually, because I think there's no random seed parameter in the Kfold class.

So, to summarize what I was thinking of:

Implementing `predict_meta_features(self, X=None)`, which returns the training meta-features if `X=None` and the "test" features if `X=X`. This could be done as a generator, so that the current `fit`method can be refactored to use `predict_meta_features(self, X=X)`

What do you think about this idea?

**Comment by takashioya** [comment]
@rasbt 
Thanks for the good improvement plan.

> Rename it to make it “public” method

OK, I got it.

> Implementing ```predict_meta_features(self, X=None)```, which returns the training meta-features if ```X=None``` and the "test" features if ```X=X```.

I like this user interface, because it’s very easy for everyone to use, and not confusing.
However, I think it has 2 problems. (To be accurate, the first one is not a problem, but just a modification)
The first one is about the modification of ```fit```. To get ‘train’ meta_feature, we need to save ```instance```.
 I showed what ‘instance’ means in Code_3: 

Code_3:
```python
def fit(self, X, y, groups=None):
    kfold = check_cv(self.cv, y)
    if isinstance(self.cv, int):
        kfold.shuffle = self.shuffle

    meta_features = np.zeros((X.shape[0], len(self.regressors)))

    for i, regr in enumerate(self.regressors):

        for train_idx, holdout_idx in kfold.split(X, y, groups):
            instance = clone(regr)
            instance.fit(X[train_idx], y[train_idx])
	    #we need to save ‘instance’ to predict ‘train’ meta features.
            y_pred = instance.predict(X[holdout_idx])
            meta_features[holdout_idx, i] = y_pred

    if self.use_features_in_secondary:
        self.meta_regr_.fit(np.hstack((X, meta_features)), y)
    else:
        self.meta_regr_.fit(meta_features, y)

    for regr in self.regr_:
        regr.fit(X, y)

    return self
```

The second one is a problem of calculation time. If we don’t save ```meta_features``` in ```fit```, we have to do completely same prediction twice.
When we use models which take more time in prediction than training (like KNN), this problem will be serious.
I showed the situation in Code_4.

Code_4:
```python
model = StackingCVRegressor()
#we already predict 'train' meta features here  
model.fit(X_train, y_train)
#we have to do same prediction here
train_meta_features = model.predict_meta_features(X_train)
```

So, I think we need parameters like ```save_meta_features(bool)``` in ```fit```, and don’t predict anything in ```predict_meta_features```. It is a bit strange to have an prediction array as StackingCVRegressor’s attributes, though.
I showed the example in Code_5.

Code_5:
```python
def fit(self, X, y, groups=None, save_meta_features = False):
    kfold = check_cv(self.cv, y)
    if isinstance(self.cv, int):
        kfold.shuffle = self.shuffle

    #This is the meta features for the training set
    meta_features = np.zeros((X.shape[0], len(self.regressors)))

    for i, regr in enumerate(self.regressors):

        for train_idx, holdout_idx in kfold.split(X, y, groups):
            instance = clone(regr)
            instance.fit(X[train_idx], y[train_idx])
            y_pred = instance.predict(X[holdout_idx])
            meta_features[holdout_idx, i] = y_pred
    if save_meta_features:
        self.train_meta_features = meta_features

    if self.use_features_in_secondary:
        self.meta_regr_.fit(np.hstack((X, meta_features)), y)
    else:
        self.meta_regr_.fit(meta_features, y)

    for regr in self.regr_:
        regr.fit(X, y)

    return self

def predict_meta_features(self, X = None):
    if X:
        return np.column_stack([regr.predict(X) for regr in self.regr_])
    else:
        #Exception handling may be needed here
        return self.train_meta_features

```
How do you think about this?

**Comment by rasbt** [comment]
Thanks for thinking about this so thoroughly, and I agree with your concerns :). Like you suggested, the best way might be to store the training meta features as an attribute. Although, it may seem a bit awkward, it's also done in the scikit-learn API.

However, instead of passing an argument to `fit`, I would suggest to do it in the constructor, which can potentially avoid other complications down the road (e.g., if someone wants to use the StackingCVRegressor as part of a scikit-learn pipeline, which calls fit without extra arguments).

So, what I am suggesting is something like

```python
StackingCVRegressor(..., store_train_meta_features=False)`
```

and the attribute would be `train_meta_features_`, similar to what's been done in scikit-learn in e.g., the `LinearDiscriminantAnalysis` with the `store_covariance` argument and `covariance_`

In addition, I would still suggest to have a method

```python
predict_meta_features(self, X)
```
to compute meta features for e.g., the test set and other datasets.

Does this sound good?

**Comment by takashioya** [comment]
@rasbt
Thanks for the good suggestion!
I completely agree with your idea and now it should be the best way. I’ll start coding to send PR :)

**Comment by rasbt** [comment]
That's great, thanks a lot!
